hedera-node/hedera-app/src/main/java/com/hedera/node/app/blocks/impl/streaming/BlockNodeConnectionManager.java

```diff
@@ -255,29 +259,204 @@ public void shutdown() {
-                        future.complete(false);
+                    if (!primaryActive() && !secondaryActive()) {
+                        // Since neither connection is active, we can try to establish a new connection
+                        future.complete(true); // Ensure future completed
+                    } else if (Instant.now().isAfter(deadline)) {
+                        future.complete(false); // Handle timeout
                     }
+                    // Otherwise we just keep waiting, don't return anything
                 },
                 0,
-                1,
-                TimeUnit.SECONDS);
+                100,
+                TimeUnit.MILLISECONDS);
+
+        try {
+            return future.get(timeout.toMillis(), TimeUnit.MILLISECONDS); // Avoid indefinite blocking
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+            Thread.currentThread().interrupt();
+            return false;
+        } finally {
+            scheduledTask.cancel(true);
+        }
+    }
 
-        return future.join();
+    static String blockNodeName(@Nullable final BlockNodeConfig node) {
+        return node != null ? node.address() + ":" + node.port() : "null";
     }
 
     /**
-     * Returns the gRPC endpoint for the block stream service.
-     * @return the gRPC endpoint
+     * Attempts to establish a connection to a block node based on priority.
      */
-    public String getGrpcEndPoint() {
-        return GRPC_END_POINT;
+    private void establishConnection() {
+        logger.info("Establishing connection to primary block node");
+        List<BlockNodeConfig> availableNodes = blockNodeConfigurations.getAllNodes();
+
+        final Map<Integer, List<BlockNodeConfig>> priorityGroups =
+                availableNodes.stream().collect(Collectors.groupingBy(BlockNodeConfig::priority));
+        final List<Integer> sortedPriorities = new ArrayList<>(priorityGroups.keySet());
+        sortedPriorities.sort(Integer::compare);
+
+        // Find the current lowest priority of active connections
+        final int currentMinPriority = getCurrentMinPriority();
+
+        BlockNodeConfig selectedNode = null;
+        for (Integer priority : sortedPriorities) {
+            // Skip over any nodes that have less priority than the current minimum
+            if (priority >= currentMinPriority) continue;
+
+            final List<BlockNodeConfig> nodesInGroup = new ArrayList<>(priorityGroups.get(priority));
+            Collections.shuffle(nodesInGroup);
+            selectedNode = nodesInGroup.getFirst();
+            if (!connectionsInRetry.containsKey(selectedNode)) {
+                break;
+            }
+        }
+
+        if (selectedNode == null) {
+            throw new IllegalStateException(
+                    "No available block node found for connection. Check configuration and network status.");
+        }
+
+        connectToNode(selectedNode);
```

Comment:
Shouldn't we try to establish secondary connection as well?

Reply:
Yes, if the first is unsuccessful we should try again. Done!